drivers/hmis/app/models/hmis/form/instance.rb

```diff
@@ -39,7 +41,7 @@ class Hmis::Form::Instance < ::GrdaWarehouseBase
                      )
                    end
 
-  scope :with_role, ->(role) { joins(:definition).where(fd_t[:role].in(role)) }
+  scope :with_role, ->(role) { joins(:definition).where(fd_t[:role].in(role).and(fd_t[:status].eq(:published))) }
```

Comment:
I'm not sure this is the right place for this change; we can also move it onto the client detail forms query if we are worried about broader impact. But I think if we don't put it here we might see this bug crop up again (where the scope returns multiple duplicate Instances because of the join to Definition)

Reply:
Yeah this makes me nervous. There are 21 usages of this scope, so we potentially impacting a lot of behavior. I would feel more comfortable auditing them each and making changes one-by-one. I can already see there are a few places where we use this scope and want it to return instances regardless of whether the form has a published version or not. (For example `form_rules` on query_type, and `assessment_names_for_project` which is supposed to include inactive/retired since its just used for filtering existing assessments).

So, I think I would prefer to add a scope `scope :published, ->(role) { joins(:definition).merge(Hmis::Form::Definition.published) }` and call it from the client detail forms query.

It would be wise to audit other places we use the instance though.. I can see a few places we might want to add the published scope, like possibly `external_form_types_for_project`. Lets split that work out from this bug fix, though.

Reply:
Created https://github.com/open-path/Green-River/issues/6793

Reply:
Thanks!